File: daily/2024/874_Walking_Robot_Simulation.cpp
#include <iostream>
#include <vector>
#include <unordered_set>

using namespace std;

class Solution {
public:
    struct pair_hash{
        template <class T1, class T2>
        size_t operator()(const pair<T1,T2>& a) const{
            int h1 = hash<T1>{}(a.first);
            int h2 = hash<T2>{}(a.second);
            return h1 ^ h2;
        }
    };
    int robotSim(vector<int>& commands, vector<vector<int>>& obstacles) {
        int ans = 0;
        unordered_set<pair<int,int>, pair_hash> s;
        for (auto& obstacle : obstacles){
            s.insert({obstacle[0], obstacle[1]});
        }
        int x = 0;
        int y = 0;
        int dir = 0;
        vector<pair<int,int>> d = {{0,1},{1,0},{0,-1},{-1,0}};
        for (int command : commands){
            if (command == -1) dir = (dir + 1) % 4;
            else if (command == -2) dir = (dir + 4 - 1) % 4;
            else{
                for (int k = 0; k < command; k++){
                    int nx = x + d[dir].first;
                    int ny = y + d[dir].second;
                    if (s.find({nx, ny}) != s.end()) continue;
                    x = nx, y = ny;
                    ans = max(ans, x * x + y * y);
                }
            }
        }
        return ans;
    }
};

int main(){
    vector<int> commands = {4,-1,3};
    vector<vector<int>> obstacles = {};
    // vector<int> commands = {4,-1,4,-2,4};
    // vector<vector<int>> obstacles = {{2,4}};
    // vector<int> commands = {6,-1,-1,6};
    // vector<vector<int>> obstacles = {};
    Solution S;
    cout << S.robotSim(commands, obstacles) << endl;
    return 0;
}

File: daily_python/874_Walking_Robot_Simulation.py
import os
import sys
from typing import List

class Solution:
    def robotSim(self, commands: List[int], obstacles: List[List[int]]) -> int:
        ans = 0
        obs = set((o[0], o[1]) for o in obstacles)
        x, y = 0, 0
        d = [[0,1],[1,0],[0,-1],[-1,0]]
        dir = 0
        for command in commands:
            if command == -1:
                dir = (dir + 1) % 4
            elif command == -2:
                dir = (dir + 3) % 4
            else:
                for _ in range(command):
                    nx = x + d[dir][0]
                    ny = y + d[dir][1]
                    if (nx, ny) not in obs:
                        x, y = nx, ny
                        ans = max(ans, x * x + y * y)
        return ans

def readDataSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            commands = list(map(int, lines[0].split('=')[1].strip()[1:-2].split(',')))
            obstacle = lines[1].split('=')[1].strip()[2:-3].split('},{')
            obstacles = [list(map(int, o.split(','))) for o in obstacle] if obstacle != [''] else []
            dataset.append((commands, obstacles))
    return dataset

if __name__ == '__main__':
    if (len(sys.argv) == 1):
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for commands, obstacles in dataset:
        results.append(solution.robotSim(commands, obstacles))
    for index, result in enumerate(results):
        print(f'Example {index + 1} : {result}')